src/gtest/test_noteencryption.cpp

```diff
@@ -1,5 +1,180 @@
+    uint256 pk_enc = ZCNoteEncryption::generate_pubkey(sk_enc);
+
+    ZCNoteEncryption b = ZCNoteEncryption(uint256());
+    for (size_t i = 0; i < 100; i++)
+    {
+        ZCNoteEncryption c = ZCNoteEncryption(uint256());
+
+        ASSERT_TRUE(b.get_epk() != c.get_epk());
+    }
+
+    boost::array<unsigned char, 216> message;
+    for (unsigned char i = 0; i < 216; i++) {
+        // Fill the message with dummy data
+        message[i] = (unsigned char) i;
+    }
+
+    for (int i = 0; i < 255; i++) {
+        auto ciphertext = b.encrypt(pk_enc, message);
+
+        {
+            ZCNoteDecryption decrypter(sk_enc);
+
+            // Test decryption
+            auto plaintext = decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i);
+            ASSERT_TRUE(plaintext == message);
+
+            // Test wrong nonce
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), (i == 0) ? 1 : (i - 1)), std::runtime_error);
+        
+            // Test wrong ephemeral key
+            {
+                ZCNoteEncryption c = ZCNoteEncryption(uint256());
+
+                ASSERT_THROW(decrypter.decrypt(ciphertext, c.get_epk(), uint256(), i), std::runtime_error);
+            }
+        
+            // Test wrong seed
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256S("11035d60bc1983e37950ce4803418a8fb33ea68d5b937ca382ecbae7564d6a77"), i), std::runtime_error);
+        
+            // Test corrupted ciphertext
+            ciphertext[10] ^= 0xff;
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i), std::runtime_error);
+            ciphertext[10] ^= 0xff;
+        }
+
+        {
+            // Test wrong private key
+            uint256 sk_enc_2 = ZCNoteEncryption::generate_privkey(uint256());
+            ZCNoteDecryption decrypter(sk_enc_2);
+
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i), std::runtime_error);
+        }
+
+        {
+            TestNoteDecryption decrypter(sk_enc);
+
+            // Test decryption
+            auto plaintext = decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i);
+            ASSERT_TRUE(plaintext == message);
+
+            // Test wrong public key (test of KDF)
+            decrypter.change_pk_enc(uint256());
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i), std::runtime_error);
+        }
+    }
+
+    // Nonce space should run out here
+    ASSERT_THROW(b.encrypt(pk_enc, message), std::logic_error);
+}
+
+uint256 test_prf(
+    unsigned char distinguisher,
+    uint256 x,
+    uint256 y
+) {
+    *x.begin() &= 0x0f;
+    *x.begin() |= distinguisher;
+    CSHA256 hasher;
+    hasher.Write(x.begin(), 32);
+    hasher.Write(y.begin(), 32);
+
+    uint256 ret;
+    hasher.FinalizeNoPadding(ret.begin());
+    return ret;
+}
+
+TEST(noteencryption, prf_addr)
```

Comment:
The PRFs aren't strictly speaking part of the NoteEncryption. Maybe they should be tested in a separate file? (Does not block this PR.)